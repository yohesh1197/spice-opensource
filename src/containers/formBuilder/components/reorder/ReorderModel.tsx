import { FormApi } from 'final-form';
import arrayMutators from 'final-form-arrays';
import { cloneDeep } from 'lodash';
import { useRef } from 'react';
import { Field, Form } from 'react-final-form';
import { FieldArray } from 'react-final-form-arrays';
import ModalViewer from '../../../../components/modal/ModalViewer';
import ReorderContainer from '../../../../components/reorder/ReorderContainer';
import ReorderItem from '../../../../components/reorder/ReorderItem';
import APPCONSTANTS from '../../../../constants/appConstants';
import { composeValidators, convertToCaptilize, required, convertToSubString } from '../../../../utils/validation';
import styles from '../../styles/FormBuilder.module.scss';
import { IFieldViewType as IViewType } from '../../types/ComponentConfig';
export const ReorderModel = ({
  initialValue,
  orderRef,
  formName,
  onSubmit,
  onCancel,
  isRemoveItem,
  reorderTitle = 'Edit Order',
  familyName
}: {
  initialValue: any;
  orderRef: any;
  formName: string;
  onSubmit: () => void;
  onCancel: () => void;
  isRemoveItem?: boolean;
  reorderTitle?: string;
  familyName?: string;
}) => {
  let formattedValue: any = null;

  if (!familyName) {
    const sortAndgroupByFamily = () => {
      const formValues = cloneDeep(initialValue.values);
      const groupedFamily: Array<{ [x: string]: any }> = [];
      Object.keys(formValues)
        .sort(
          (fieldA: string, fieldB: string) =>
            (formValues[fieldA][fieldA]?.familyOrder || 0) - (formValues[fieldB][fieldB]?.familyOrder || 0)
        )
        .forEach((groupedFamilyName: string) => {
          if (groupedFamilyName !== APPCONSTANTS.NO_FAMILY) {
            groupedFamily.push(formValues[groupedFamilyName][groupedFamilyName]);
          }
        });
      return groupedFamily;
    };
    formattedValue = sortAndgroupByFamily();
  } else {
    const formValues = cloneDeep(initialValue.values);
    const currentFamilyGroup = formValues[familyName];
    // remove card view object
    delete currentFamilyGroup[familyName];
    formattedValue = Object.values(currentFamilyGroup).sort((a: any, b: any) => a.orderId - b.orderId);
  }

  const checkDuplication = (value: any, values: any, i: number) => {
    let duplicateIndex: number | undefined;
    if (
      values[formName]?.find((res: any, index: number) => {
        if (res?.name === value && i !== index) {
          duplicateIndex = index;
          return true;
        }
        return false;
      })
    ) {
      return `duplicate_${duplicateIndex}_${i}`;
    }
  };

  const formRef = useRef<FormApi<any>>();
  const idRefs = useRef<string[]>(
    (() => (formattedValue ? formattedValue.map((item: any) => item.id as string) : []))()
  );
  const onReorder = (order: { [itemId: string]: number }) => {
    orderRef.current = { ...order };
    idRefs.current.forEach((itemId: string, i: number) => {
      formRef.current?.change(`${formName}[${i}].displayOrder` as keyof IViewType, String(order[itemId]));
    });
  };

  return (
    <ModalViewer show={true} title={reorderTitle} handleCancel={onCancel} size='modal-md'>
      <Form
        onSubmit={() => undefined}
        mutators={{
          ...arrayMutators
        }}
        initialValues={{ is_active: true }}
        render={({ form }) => {
          formRef.current = form;
          return (
            <>
              <div className={`modal-body ${styles.scroll}`}>
                <FieldArray name={formName} initialValue={formattedValue}>
                  {({ fields }) => {
                    return (
                      <ReorderContainer onReorder={onReorder}>
                        {fields.map((name, index) => {
                          if (!idRefs.current[index]) {
                            idRefs.current[index] = fields.value[index].id as string;
                          }
                          return (
                            <ReorderItem
                              initOrder={index}
                              itemId={idRefs.current[index].toString()}
                              isRemoveItem={isRemoveItem}
                              key={idRefs.current[index].toString()}
                            >
                              <Field
                                name={`${name}.name`}
                                type='text'
                                validate={composeValidators(required, (value: string, values: any) =>
                                  checkDuplication(value, values, index)
                                )}
                                format={convertToCaptilize}
                                parse={convertToCaptilize}
                                render={() => {
                                  return (
                                    <div className={`${styles.customizationFont}`}>
                                      {convertToSubString(fields.value[index].title)}
                                    </div>
                                  );
                                }}
                              />
                              <Field name={`${name}.displayOrder`} initialValue={index + 1} render={() => null} />
                            </ReorderItem>
                          );
                        })}
                      </ReorderContainer>
                    );
                  }}
                </FieldArray>
              </div>
              <div className='modal-footer py-0dot75 px-1dot25'>
                <button type='button' className='btn secondary-btn me-0dot625 px-1dot125 ms-auto' onClick={onCancel}>
                  Cancel
                </button>
                <button type='button' className='btn primary-btn px-1dot75' onClick={onSubmit}>
                  Confirm
                </button>
              </div>
            </>
          );
        }}
      />
    </ModalViewer>
  );
};
