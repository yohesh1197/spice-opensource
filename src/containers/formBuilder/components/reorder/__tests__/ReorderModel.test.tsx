import { mount } from 'enzyme';
import React from 'react';
import { ReorderModel } from '../ReorderModel';
import MOCK_DATA_CONSTANTS from '../../../../../components/reorder/ReorderConstants'

describe('ReorderModel component', () => {
  const onSubmit = jest.fn();
  const onCancel = jest.fn();
  const orderRef = { current: {} };
  const formName = 'fields';
  const initialValue = {
    values: MOCK_DATA_CONSTANTS.INITIAL_VALUES
  };

  it('should render correctly', () => {
    const wrapper = mount(
      <ReorderModel
        initialValue={initialValue}
        orderRef={orderRef}
        formName={formName}
        onSubmit={onSubmit}
        onCancel={onCancel}
      />
    );
    wrapper.setProps({formattedValue: [{id: 1}, {id: 2}]})
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onSubmit when Confirm button is clicked', () => {
    const wrapper = mount(
      <ReorderModel
        initialValue={initialValue}
        orderRef={orderRef}
        formName={formName}
        onSubmit={onSubmit}
        onCancel={onCancel}
      />
    );

    wrapper.find('.primary-btn').simulate('click');

    expect(onSubmit).toHaveBeenCalled();
  });

  it('should call onCancel when Cancel button is clicked', () => {
    const wrapper = mount(
      <ReorderModel
        initialValue={initialValue}
        orderRef={orderRef}
        formName={formName}
        onSubmit={onSubmit}
        onCancel={onCancel}
      />
    );

    wrapper.find('.secondary-btn').simulate('click');

    expect(onCancel).toHaveBeenCalled();
  });
  
  it('should render when family name is given', () => {
    const wrapper = mount(
      <ReorderModel
        initialValue={initialValue}
        orderRef={orderRef}
        formName={formName}
        onSubmit={onSubmit}
        onCancel={onCancel}
        familyName='bioData'
      />
    );
    wrapper.setProps({formattedValue: [{id: 1}, {id: 2}]})
    expect(wrapper).toMatchSnapshot();

    const ReorderContainer = wrapper.find('ReorderContainer');
    const onReorder:any = ReorderContainer.prop('onReorder');
    onReorder({bioData: 2, bioMetrics: 0, bpLog: 1, glucoseLog: 3, phq4: 4});
  });
});
