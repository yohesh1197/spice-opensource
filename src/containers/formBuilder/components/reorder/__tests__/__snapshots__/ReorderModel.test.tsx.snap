// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReorderModel component should render correctly 1`] = `ReactWrapper {}`;

exports[`ReorderModel component should render when family name is given 1`] = `ReactWrapper {}`;
